File: apps/api/src/queue/sqs-queue-client.ts
/**
 * SQS backend — the AWS path (Requirement 5.3).
 *
 * | Operation    | SQS API                                                     |
 * |--------------|-------------------------------------------------------------|
 * | `enqueue`    | `SendMessage`                                               |
 * | `receive`    | `ReceiveMessage` with long polling                          |
 * | `ack`        | `DeleteMessage`                                             |
 * | `deadLetter` | `SendMessage` to the DLQ + `DeleteMessage`, or the redrive policy |
 * | `depth`      | `GetQueueAttributes ApproximateNumberOfMessages`            |
 *
 * The message body is the same JSON text the Redis backend stores, and no
 * contract data is carried in message attributes, so a message captured from
 * one backend can be replayed into the other.
 */

import { describeJob, parsePublishJob, serializePublishJob } from './publish-job.js';
import type {
  DeadLetterEvent,
  JobHandle,
  PublishJob,
  QueueClient,
  ReceivedJob,
} from './types.js';

/** SQS caps `WaitTimeSeconds` at 20. */
export const SQS_MAX_WAIT_SECONDS = 20;

export interface SqsMessage {
  readonly messageId: string;
  readonly receiptHandle: string;
  readonly body: string;
}

/**
 * The narrow slice of SQS this client uses. `AwsSqsPort` in `aws-sqs-port.ts`
 * implements it over `@aws-sdk/client-sqs`; the unit tests implement it with a
 * fake, so no test needs AWS credentials.
 */
export interface SqsPort {
  sendMessage(input: {
    queueUrl: string;
    messageBody: string;
    messageAttributes?: Record<string, string> | undefined;
  }): Promise<void>;
  receiveMessage(input: { queueUrl: string; waitTimeSeconds: number }): Promise<SqsMessage[]>;
  deleteMessage(input: { queueUrl: string; receiptHandle: string }): Promise<void>;
  approximateNumberOfMessages(queueUrl: string): Promise<number>;
  close(): Promise<void>;
}

export interface SqsQueueClientOptions {
  queueUrl: string;
  /**
   * Optional. When set, `deadLetter` sends the message to this queue explicitly
   * and deletes it from the main queue. When unset, the message is left in place
   * for the queue's redrive policy to move after `maxReceiveCount` receives.
   */
  deadLetterQueueUrl?: string | undefined;
  onDeadLetter?: ((event: DeadLetterEvent) => void) | undefined;
}

export class SqsQueueClient implements QueueClient {
  private readonly sqs: SqsPort;
  private readonly queueUrl: string;
  private readonly deadLetterQueueUrl: string | null;
  private readonly onDeadLetter: (event: DeadLetterEvent) => void;

  constructor(sqs: SqsPort, options: SqsQueueClientOptions) {
    this.sqs = sqs;
    this.queueUrl = options.queueUrl;
    this.deadLetterQueueUrl = options.deadLetterQueueUrl ?? null;
    this.onDeadLetter = options.onDeadLetter ?? (() => {});
  }

  async enqueue(job: PublishJob): Promise<void> {
    await this.sqs.sendMessage({
      queueUrl: this.queueUrl,
      messageBody: serializePublishJob(job),
    });
  }

  async receive(waitSeconds: number): Promise<ReceivedJob | null> {
    const waitTimeSeconds = Math.min(
      SQS_MAX_WAIT_SECONDS,
      Math.max(0, Math.trunc(waitSeconds)),
    );

    const messages = await this.sqs.receiveMessage({
      queueUrl: this.queueUrl,
      waitTimeSeconds,
    });

    const message = messages[0];
    if (message === undefined) {
      return null;
    }

    const handle: JobHandle = {
      backend: 'sqs',
      messageId: message.messageId,
      receiptHandle: message.receiptHandle,
    };
    const parsed = parsePublishJob(message.body);

    return parsed.ok
      ? { raw: message.body, handle, job: parsed.job, invalidReason: null, invalidDetail: null }
      : {
          raw: message.body,
          handle,
          job: null,
          invalidReason: parsed.reason,
          invalidDetail: parsed.detail,
        };
  }

  async ack(job: ReceivedJob): Promise<void> {
    await this.sqs.deleteMessage({
      queueUrl: this.queueUrl,
      receiptHandle: this.receiptHandle(job),
    });
  }

  async deadLetter(job: ReceivedJob, reason: string): Promise<void> {
    const receiptHandle = this.receiptHandle(job);
    const { jobId, postId, attempt } = describeJob(job.job);

    if (this.deadLetterQueueUrl !== null) {
      // Body unchanged so the dead letter stays replayable; the reason rides
      // along as an attribute, which no consumer reads as contract data.
      await this.sqs.sendMessage({
        queueUrl: this.deadLetterQueueUrl,
        messageBody: job.raw,
        messageAttributes: { DeadLetterReason: reason },
      });
      await this.sqs.deleteMessage({ queueUrl: this.queueUrl, receiptHandle });
    }
    // Otherwise the message is deliberately left untouched: it becomes visible
    // again after the visibility timeout and the queue's redrive policy moves it
    // to the DLQ once `maxReceiveCount` is reached. Deleting it here would
    // discard it instead of dead-lettering it.

    this.onDeadLetter({
      backend: 'sqs',
      reason,
      jobId,
      postId,
      attempt,
      viaRedrivePolicy: this.deadLetterQueueUrl === null,
    });
  }

  async depth(): Promise<number> {
    return this.sqs.approximateNumberOfMessages(this.queueUrl);
  }

  async close(): Promise<void> {
    await this.sqs.close();
  }

  private receiptHandle(job: ReceivedJob): string {
    if (job.handle.backend !== 'sqs') {
      throw new TypeError(
        `SqsQueueClient received a "${job.handle.backend}" handle; handles are not portable across backends`,
      );
    }
    return job.handle.receiptHandle;
  }
}

File: apps/worker/src/publishhub_worker/queue/sqs_queue_client.py
"""
SQS backend — the AWS path (Requirement 5.3).

| Operation     | SQS API                                                           |
|---------------|-------------------------------------------------------------------|
| `enqueue`     | `SendMessage`                                                     |
| `receive`     | `ReceiveMessage` with long polling                                |
| `ack`         | `DeleteMessage`                                                   |
| `dead_letter` | `SendMessage` to the DLQ + `DeleteMessage`, or the redrive policy  |
| `depth`       | `GetQueueAttributes ApproximateNumberOfMessages`                  |

The message body is the same JSON text the Redis backend stores, and no contract
data is carried in message attributes, so a message captured from one backend can
be replayed into the other.

There is no reaper here: SQS has a visibility timeout of its own, so a message
claimed by a worker that dies becomes visible again without help.
"""

from __future__ import annotations

import math
from collections.abc import Mapping
from dataclasses import dataclass
from typing import Protocol

from .publish_job import describe_job, parse_publish_job, serialize_publish_job
from .types import (
    DeadLetterEvent,
    DeadLetterListener,
    PublishJob,
    ReceivedJob,
    SqsJobHandle,
)

#: SQS caps `WaitTimeSeconds` at 20.
SQS_MAX_WAIT_SECONDS = 20


@dataclass(frozen=True, slots=True)
class SqsMessage:
    message_id: str
    receipt_handle: str
    body: str


class SqsPort(Protocol):
    """
    The narrow slice of SQS this client uses. `AwsSqsPort` in `aws_sqs_port.py`
    implements it over `boto3`; the unit tests implement it with a fake, so no
    test needs AWS credentials.
    """

    def send_message(
        self,
        *,
        queue_url: str,
        message_body: str,
        message_attributes: Mapping[str, str] | None = None,
    ) -> None: ...

    def receive_message(self, *, queue_url: str, wait_time_seconds: int) -> list[SqsMessage]: ...

    def delete_message(self, *, queue_url: str, receipt_handle: str) -> None: ...

    def approximate_number_of_messages(self, queue_url: str) -> int: ...

    def close(self) -> None: ...


class SqsQueueClient:
    """
    SQS implementation of `QueueClient`.

    `dead_letter_queue_url` is optional. When set, `dead_letter` sends the message
    to that queue explicitly and deletes it from the main queue. When unset, the
    message is left in place for the queue's redrive policy to move after
    `maxReceiveCount` receives.
    """

    def __init__(
        self,
        sqs: SqsPort,
        *,
        queue_url: str,
        dead_letter_queue_url: str | None = None,
        on_dead_letter: DeadLetterListener | None = None,
    ) -> None:
        self._sqs = sqs
        self._queue_url = queue_url
        self._dead_letter_queue_url = dead_letter_queue_url
        self._on_dead_letter = on_dead_letter

    def enqueue(self, job: PublishJob) -> None:
        self._sqs.send_message(
            queue_url=self._queue_url,
            message_body=serialize_publish_job(job),
        )

    def receive(self, wait_seconds: float) -> ReceivedJob | None:
        wait_time_seconds = min(SQS_MAX_WAIT_SECONDS, max(0, math.trunc(wait_seconds)))

        messages = self._sqs.receive_message(
            queue_url=self._queue_url,
            wait_time_seconds=wait_time_seconds,
        )
        if not messages:
            return None

        message = messages[0]
        handle = SqsJobHandle(
            message_id=message.message_id,
            receipt_handle=message.receipt_handle,
        )
        parsed = parse_publish_job(message.body)

        if parsed.ok:
            return ReceivedJob(raw=message.body, handle=handle, job=parsed.job)
        return ReceivedJob(
            raw=message.body,
            handle=handle,
            job=None,
            invalid_reason=parsed.reason,
            invalid_detail=parsed.detail,
        )

    def ack(self, job: ReceivedJob) -> None:
        self._sqs.delete_message(
            queue_url=self._queue_url,
            receipt_handle=self._receipt_handle(job),
        )

    def dead_letter(self, job: ReceivedJob, reason: str) -> None:
        receipt_handle = self._receipt_handle(job)
        described = describe_job(job.job)

        if self._dead_letter_queue_url is not None:
            # Body unchanged so the dead letter stays replayable; the reason rides
            # along as an attribute, which no consumer reads as contract data.
            self._sqs.send_message(
                queue_url=self._dead_letter_queue_url,
                message_body=job.raw,
                message_attributes={"DeadLetterReason": reason},
            )
            self._sqs.delete_message(queue_url=self._queue_url, receipt_handle=receipt_handle)
        # Otherwise the message is deliberately left untouched: it becomes visible
        # again after the visibility timeout and the queue's redrive policy moves
        # it to the DLQ once `maxReceiveCount` is reached. Deleting it here would
        # discard it instead of dead-lettering it.

        if self._on_dead_letter is not None:
            self._on_dead_letter(
                DeadLetterEvent(
                    backend="sqs",
                    reason=reason,
                    job_id=described.job_id,
                    post_id=described.post_id,
                    attempt=described.attempt,
                    via_redrive_policy=self._dead_letter_queue_url is None,
                )
            )

    def depth(self) -> int:
        return self._sqs.approximate_number_of_messages(self._queue_url)

    def close(self) -> None:
        self._sqs.close()

    def _receipt_handle(self, job: ReceivedJob) -> str:
        if job.handle.backend != "sqs":
            raise TypeError(
                f'SqsQueueClient received a "{job.handle.backend}" handle; '
                "handles are not portable across backends"
            )
        return job.handle.receipt_handle
